### PR TITLE
WatchService Provides File Name Instead of Relative Path

### PR DESCRIPTION
**Pull request by dstivanov**
The file path, passed to the AbstractWatchQueueReader.processWatchEvent is now full (till now it was containing only the name of the target file, but not its relative to the watched directory).
Added tests for the AbstractWatchService to verify the behavior above.

**Why do we need the full path**:
When implementing the AbstractWatchService, we can provide an option to track the sub-directories by implementing the AbstractWatchService.watchSubDirectories() method. However, when the sub-directories are tracked, the AbstractWatchService.registerDirectory(Path subDir) will be called for each sub-directory. 
So we have two options - either to implement separate watch service for each of the sub-directories or to have the relative path as an argument when a watch event is detected. With this PR I assume that the second option is more convenient.

Signed-off-by: Dimitar Ivanov dstivanov@gmail.com

### CODE REVIEW

**Comment by maggu2810** [comment]
After reading your explanation I think that makes sense.
I assume you have tested that this does not break any current usage.

File: bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/service/AbstractWatchServiceTest.groovy [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        boolean isDeletedWithPrecedingModify = osSpecific && kind.equals(ENTRY_DELETE) && relativeService.allFullEvents.size() == 2 && relativeService.allFullEvents[0].eventKind.equals(ENTRY_MODIFY)
+        if(isDeletedWithPrecedingModify){
+            // Remove the ENTRY_MODIFY element as it is not needed
+            relativeService.allFullEvents.remove(0)
+        }
+
+        waitForAssert {assertThat "Only one event for file $fileName is expected shortly after the file has been altered. Here are the found events: " + relativeService.allFullEvents, relativeService.allFullEvents.size(), is(1)}
+        FullEvent fullEvent = relativeService.allFullEvents[0]
+
+        assertThat "The path of the processed event should be relative to the root", fullEvent.eventPath.toString(), is(ROOT + File.separatorChar + fileName)
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/service/AbstractWatchServiceTest.groovy]: I do not understand this assert.
If the path is expected to be relative to ROOT, it should be `fileName`, but you are checking that it is `ROOT + File.separatorChar + fileName`.

**Comment by kaikreuzer** [comment]
> Why do we need the full path

Does full path mean an absolute path or a path relative to the watched folder?
Please make sure that you do not introduce any regression. For folders that are watched without any subfolder (as done today), the path should continue to be simply the file name, because this is what current implementations expect (and I would not see any reason why this should be changed).

File: bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/service/AbstractWatchServiceTest.groovy [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        sleep 1000
+
+        // As the implementation of the watch events is OS specific, it can happen that when the file is deleted two events are fired - ENTRY_MODIFY followed by an ENTRY_DELETE
+        // This is usually observed on Windows and below is the workaround
+        // Related discussion in StackOverflow: http://stackoverflow.com/questions/28201283/watchservice-windows-7-when-deleting-a-file-it-fires-both-entry-modify-and-e
+        boolean isDeletedWithPrecedingModify = osSpecific && kind.equals(ENTRY_DELETE) && relativeService.allFullEvents.size() == 2 && relativeService.allFullEvents[0].eventKind.equals(ENTRY_MODIFY)
+        if(isDeletedWithPrecedingModify){
+            // Remove the ENTRY_MODIFY element as it is not needed
+            relativeService.allFullEvents.remove(0)
+        }
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/service/AbstractWatchServiceTest.groovy]: Can't we handle the above logic in a way, that we do not need to the "sleep 1000" at all.
It seems to be a little timing dependent.

If we wait for a special event, I assume we can simply wait for that event and ignore all others. Can't we?

Comment (dstivanov) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/service/AbstractWatchServiceTest.groovy]: Good point! I will add a wait for assert here with a 5 seconds timeout. So if no event is detected within the 5 seconds timeout the test will fail. Thank you for the remark!

File: bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/service/AbstractWatchServiceTest.groovy (hunk 1 above)
Comment (dstivanov) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/service/AbstractWatchServiceTest.groovy]: You are completely right, the path now is not relative, but absolute. I found out that I have to do some refactoring in order to make it build the relative path, so soon I will update the PR and also add some more tests for testing the relative path accordingly. Thanks for the remark!

**Comment by dstivanov** [comment]
@maggu2810 In regard to your first comment on the PR - I checked the functionalities, relying on the watch service and can not see any problems (at lease logically). Unfortunately, these usages are not covered by tests so I will try to consider writing some. Taking into account the follow up comments from Kai and you I will add some adaptation on the code in order to make the path really relative, but not absolute as it is now.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchService.java [hunk 3]
```diff
@@ -127,14 +145,20 @@ protected void stopWatchService() {
...
     protected abstract boolean watchSubDirectories();
 
     /**
-     * @param subDir
-     * @throws IOException
+     * Registers a directory to be watched by the watch service. The {@link WatchKey} of the registration should be
+     * provided.
+     * 
+     * @param directory - the directory, which will be registered in the watch service
+     * @return The {@link WatchKey} of the registration or <code>null</code> if not registration has been done.
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchService.java]: not => no

**Comment by maggu2810** [comment]
JavaDoc: please remove the dash between the parameter name and the description.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchQueueReader.java [hunk 4]
```diff
@@ -49,15 +55,45 @@
     }
 
     /**
+     * Build the object with the given parameters. The directory changes will be watched by default, e.g.
+     * watchingDirectoryChanges will be set to <code>true</code> (see {@link #setWatchingDirectoryChanges(boolean)})
+     * 
+     * @param watchService
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchQueueReader.java]: line break is not necessary

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchQueueReader.java [hunk 5]
```diff
@@ -116,4 +182,27 @@ public void run() {
...
+    /**
+     * If the queue reader is watching the directory changes, all the watch events will be processed. Otherwise the
+     * events for changed directories will be skipped. For example, on some platforms an event for modified directory is
+     * generated when a new file is created within the directory. However, this behavior could vary a lot, depending on
+     * the platform (for more information see "Platform dependencies" section in the {@link WatchService} documentation)
+     * 
+     * @param watchDirectoryChanges - set to <code>true</code> if the directory events have to be processed and
+     *            <code>false</code> otherwise
+     */
+    public void setWatchingDirectoryChanges(boolean watchDirectoryChanges) {
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchQueueReader.java]: Functions that are called in the constructor should be use the final keyword.
Without the final keyword a derived class could overwrite it and could stuff before the parent class is fully initialized.

**Comment by maggu2810** [comment]
I am still a little bit irritated that you are using different documentation style in your PR.
In the AbstractWatchQueueReader you have sometimes empty lines between the @param tags and sometimes you begin a new line after the parameter name (after @param).
Is this copied by the old code?
Github shows that as inserted / new lines of code.

About the code itself, I am fine with (at least you added some tests).
Let's wait for a last review by @kaikreuzer

**Comment by dstivanov** [comment]
Hi @maggu2810 

Yes, some parts of the javadoc are left from the old version of the code. I agree with you and will make the javadoc as consistent as I can according to your comments.

Regards,
Dimitar

**Comment by kaikreuzer** [comment]
Seems to work, thanks!

**Comment by kaikreuzer** [comment]
FTR, I reverted this PR with https://github.com/eclipse/smarthome/pull/1755.

**Comment by kaikreuzer** [comment]
It breaks the implementation of the FolderObserver, see this log excerpt:

```
19:31:02.754 [WARN ] [.core.internal.folder.FolderObserver] - Cannot open file '/Users/kai/Downloads/oh/conf/items/items/demo.items' for reading.
java.io.FileNotFoundException: File '/Users/kai/Downloads/oh/conf/items/items/demo.items' does not exist
	at org.apache.commons.io.FileUtils.openInputStream(FileUtils.java:254)[30:org.apache.commons.io:2.2.0]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.checkFile(FolderObserver.java:245)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.access$1(FolderObserver.java:240)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver$WatchQueueReader.processWatchEvent(FolderObserver.java:122)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.core.service.AbstractWatchQueueReader.run(AbstractWatchQueueReader.java:125)[96:org.eclipse.smarthome.core:0.9.0.201606271451]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_40]
19:31:02.754 [WARN ] [.core.internal.folder.FolderObserver] - Cannot open file '/Users/kai/Downloads/oh/conf/persistence/persistence/logging.persist' for reading.
java.io.FileNotFoundException: File '/Users/kai/Downloads/oh/conf/persistence/persistence/logging.persist' does not exist
	at org.apache.commons.io.FileUtils.openInputStream(FileUtils.java:254)[30:org.apache.commons.io:2.2.0]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.checkFile(FolderObserver.java:245)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.access$1(FolderObserver.java:240)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver$WatchQueueReader.processWatchEvent(FolderObserver.java:122)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.core.service.AbstractWatchQueueReader.run(AbstractWatchQueueReader.java:125)[96:org.eclipse.smarthome.core:0.9.0.201606271451]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_40]
19:31:02.755 [WARN ] [.core.internal.folder.FolderObserver] - Cannot open file '/Users/kai/Downloads/oh/conf/persistence/persistence/rrd4j.persist' for reading.
java.io.FileNotFoundException: File '/Users/kai/Downloads/oh/conf/persistence/persistence/rrd4j.persist' does not exist
	at org.apache.commons.io.FileUtils.openInputStream(FileUtils.java:254)[30:org.apache.commons.io:2.2.0]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.checkFile(FolderObserver.java:245)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.access$1(FolderObserver.java:240)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver$WatchQueueReader.processWatchEvent(FolderObserver.java:122)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.core.service.AbstractWatchQueueReader.run(AbstractWatchQueueReader.java:125)[96:org.eclipse.smarthome.core:0.9.0.201606271451]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_40]
19:31:02.755 [WARN ] [.core.internal.folder.FolderObserver] - Cannot open file '/Users/kai/Downloads/oh/conf/rules/rules/demo.rules' for reading.
java.io.FileNotFoundException: File '/Users/kai/Downloads/oh/conf/rules/rules/demo.rules' does not exist
	at org.apache.commons.io.FileUtils.openInputStream(FileUtils.java:254)[30:org.apache.commons.io:2.2.0]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.checkFile(FolderObserver.java:245)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.access$1(FolderObserver.java:240)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver$WatchQueueReader.processWatchEvent(FolderObserver.java:122)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.core.service.AbstractWatchQueueReader.run(AbstractWatchQueueReader.java:125)[96:org.eclipse.smarthome.core:0.9.0.201606271451]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_40]
19:31:02.756 [WARN ] [.core.internal.folder.FolderObserver] - Cannot open file '/Users/kai/Downloads/oh/conf/scripts/scripts/demo.script' for reading.
java.io.FileNotFoundException: File '/Users/kai/Downloads/oh/conf/scripts/scripts/demo.script' does not exist
	at org.apache.commons.io.FileUtils.openInputStream(FileUtils.java:254)[30:org.apache.commons.io:2.2.0]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.checkFile(FolderObserver.java:245)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.access$1(FolderObserver.java:240)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver$WatchQueueReader.processWatchEvent(FolderObserver.java:122)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.core.service.AbstractWatchQueueReader.run(AbstractWatchQueueReader.java:125)[96:org.eclipse.smarthome.core:0.9.0.201606271451]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_40]
19:31:02.757 [WARN ] [.core.internal.folder.FolderObserver] - Cannot open file '/Users/kai/Downloads/oh/conf/sitemaps/sitemaps/demo.sitemap' for reading.
java.io.FileNotFoundException: File '/Users/kai/Downloads/oh/conf/sitemaps/sitemaps/demo.sitemap' does not exist
	at org.apache.commons.io.FileUtils.openInputStream(FileUtils.java:254)[30:org.apache.commons.io:2.2.0]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.checkFile(FolderObserver.java:245)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.access$1(FolderObserver.java:240)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver$WatchQueueReader.processWatchEvent(FolderObserver.java:122)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.core.service.AbstractWatchQueueReader.run(AbstractWatchQueueReader.java:125)[96:org.eclipse.smarthome.core:0.9.0.201606271451]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_40]
19:31:02.757 [WARN ] [.core.internal.folder.FolderObserver] - Cannot open file '/Users/kai/Downloads/oh/conf/things/things/demo.things' for reading.
java.io.FileNotFoundException: File '/Users/kai/Downloads/oh/conf/things/things/demo.things' does not exist
	at org.apache.commons.io.FileUtils.openInputStream(FileUtils.java:254)[30:org.apache.commons.io:2.2.0]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.checkFile(FolderObserver.java:245)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.access$1(FolderObserver.java:240)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver$WatchQueueReader.processWatchEvent(FolderObserver.java:122)[115:org.eclipse.smarthome.model.core:0.9.0.201606271451]
	at org.eclipse.smarthome.core.service.AbstractWatchQueueReader.run(AbstractWatchQueueReader.java:125)[96:org.eclipse.smarthome.core:0.9.0.201606271451]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_40]
```

**Pull request merged**